Patch add str_starts_with and str_ends_with

wkhudgins92 opened the pull request:
This is the patch for the following RFC: https://wiki.php.net/rfc/add_str_starts_with_and_ends_with_functions

Contains the str_starts_with and str_ends_with functions as well as tests.

Overall this is a narrowed-down revival of https://wiki.php.net/rfc/add_str_begin_and_end_functions and #2049 .

glensc commented:
@wkhudgins92 [the wiki](https://wiki.php.net/rfc/add_str_starts_with_and_ends_with_functions) states that php8 has been released, can you link that text or correct it. as to my knowledge php8 is not yet released (to be released in December 2020):
> PHP8 has now been released and several individuals have requested the str_starts_with and str_ends_with functionality again. This is a simple but highly desired functionality for PHP.

nikic commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,44 @@ PHP_FUNCTION(str_contains)
+
+    for (i = haystack->len - 1, j = needle->len - 1; j >= 0; i--, j--) 
+        if (haystack->val[i] != needle->val[j]) 
+            RETURN_FALSE;
```
This should also perform a `memcmp()`, just on the end of the string.

wkhudgins92 replied:
Thanks @nikic ! I replaced it with a call to `memcmp`, let me know if you see any issues with how I did it.

guilliamxavier commented on ext/standard/basic_functions.stub.php:
```diff
@@ -588,6 +588,10 @@ function strrchr(string $haystack, string $needle): string|false {}
 
 function str_contains(string $haystack, string $needle): bool {}
 
+function str_begins_with(string $haystack, string $needle): bool {}
```
s/begins/starts/ ;)

wkhudgins92 replied:
Good catch!

guilliamxavier commented:
> Contains the str_starts_with and str_ends_with functions **as well as tests**.

Did you forget to commit the tests? 😅 (also please include tests for empty strings and NUL byte, like https://github.com/php/php-src/pull/5179/files#diff-efcf36fef88a923719524d7a5c568689)

wkhudgins92 commented:
> > Contains the str_starts_with and str_ends_with functions **as well as tests**.
> 
> Did you forget to commit the tests? 😅 (also please include tests for empty strings and NUL byte, like https://github.com/php/php-src/pull/5179/files#diff-efcf36fef88a923719524d7a5c568689)

Ooops I forgot to commit the tests...I was really rushing to get this out last night, I knew if I didn't get it raise up last night it might take me another week to do anything on it! These are the tests from the **_old_** RFC I raised, they don't include empty strings or NULL byte but I can add those.

wkhudgins92 commented:
> @wkhudgins92 [the wiki](https://wiki.php.net/rfc/add_str_starts_with_and_ends_with_functions) states that php8 has been released, can you link that text or correct it. as to my knowledge php8 is not yet released (to be released in December 2020):
> 
> > PHP8 has now been released and several individuals have requested the str_starts_with and str_ends_with functionality again. This is a simple but highly desired functionality for PHP.

Fixed that this morning, thanks.

guilliamxavier commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,44 @@ PHP_FUNCTION(str_contains)
 }
 /* }}} */
 
+/* {{{ proto boolean str_starts_with(string haystack, string needle)
```
s/boolean/bool/ (like everywhere else in the file)
Same for str_ends_with below

guilliamxavier commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,44 @@ PHP_FUNCTION(str_contains)
 /* }}} */
 
+/* {{{ proto boolean str_starts_with(string haystack, string needle)
+   Checks if haystack strats with needle */
```
Typo "strats"
Also, to be consistent with the existing surrounding functions, maybe rather "Checks if _a string_ starts with _another_"? (and similarly for str_ends_with below)

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,19 @@
+str_ends_with() function - unit tests for str_ends_with()
+--FILE--
+<?php
+/* Prototype: boolean str_ends_with (string $haystack string $needle );
```
s/boolean/bool/ and missing comma between params
Same in str_starts_with.phpt

guilliamxavier commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,44 @@ PHP_FUNCTION(str_contains)
+        Z_PARAM_STR(needle)
+    ZEND_PARSE_PARAMETERS_END();
+
+    if (needle->len > haystack->len) {
```
Everywhere else uses the `ZSTR_LEN()`  macro rather than direct `->len` access

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,19 @@
+str_ends_with() function - unit tests for str_ends_with()
+--FILE--
+<?php
+/* Prototype: bool str_ends_with (string $haystack string $needle );
```
sorry but still missing a `,` just after `$haystack` (here and in `str_starts_with.phpt` too)
and also there's an extra space between `$needle` and `);` (here only)

wkhudgins92 commented:
@guilliamxavier I added to the test cases as you requested and fixed the typos you found. To be honest, I'm not 100% sure I did the NULL BYTE tests properly, but that was my best guess at the time. I tested hour Python and Java handle startswith and endswith functionality, and at least with empty strings it is the same. I'll get to @nikic idea of using `memcmp` for str_ends_with soon.

guilliamxavier commented:
Thank you =) The idea of empty strings tests is basically that both `("a", "")` and `("", "")` give `true` while `("", "a")` gives `false` (for both `str_{starts,ends}_with` like for `str_contains`), so the ones you added seem OK. Concerning NUL byte tests, the goal is to ensure that the functions are "binary safe" in that they handle `\0` just like any other char (and don't crash nor stop at it), e.g. that `str_starts_with` gives `true` for `("a\x00bc", "a\x00b")` (no crash) but `false` for all of `("a\x00b", "a\x00d")`, `("a\x00b", "z\x00b")`, `("a", "a\x00")` and `("a", "\x00a")` (no stop), and similarly for `str_ends_with` (with reversed strings), maybe as a complement of those you added.

wkhudgins92 commented:
@guilliamxavier I added some more to the null byte test cases. All my test seem to be passing its just one of the Travis builds fails with this error:

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated
```
Its failed like that at least two times in a row, so not sure what is going on.

I also changed `str_ends_with` to utilize a call to `memcmp` as suggested by @nikic , I believe I got that right. Unless other things come up on the PR, I'm going to incorporate some feedback I've received into the RFC page next.

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,45 @@
+var_dump(str_ends_with("", " "));
+var_dump(str_ends_with("\x00", ""));
+var_dump(str_ends_with("\x00", "\x00"));
+var_dump(str_ends_with("a", "\x00"));
```
This one should rather be `("a\x00", "\x00")` and expect `bool(true)`

wkhudgins92 replied:
This test was to make sure that the `\x00` isn't treated like an empty string. I kept this test but added the one you suggested

guilliamxavier replied:
There is already a test for `($testStr, "\x00")` (actually, there is only in the other file, but there should be in this file too, see my comment above), so `("a", "\x00")` doesn't add anything

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,45 @@
+var_dump(str_ends_with("a", "\x00"));
+var_dump(str_ends_with("a\x00b", "\x00"));
+var_dump(str_ends_with("a\x00b", "b"));
+var_dump(str_ends_with("a\x00bc", "\x00bc"));
```
I would replace those 3 tests with a single `("ab\x00c", "b\x00c")` (expect `bool(true)`)

guilliamxavier commented on ext/standard/tests/strings/str_starts_with.phpt:
```diff
@@ -0,0 +1,47 @@
+var_dump(str_starts_with("\x00a", "\x00"));
+var_dump(str_starts_with("\x00a", "a"));
+var_dump(str_starts_with("\x00ab", "\x00a"));
+var_dump(str_starts_with("\x00ab", "\x00d"));
```
I would replace the 3 tests above with a single `("a\x00bc", "a\x00b")` (expect `bool(true)`)

guilliamxavier commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,42 @@ PHP_FUNCTION(str_contains)
+    }
+
+    k = ZSTR_LEN(haystack) - ZSTR_LEN(needle);
+    RETURN_BOOL(memcmp(&(ZSTR_VAL(haystack))[k], ZSTR_VAL(needle), ZSTR_LEN(needle)) == 0);
```
`&(ZSTR_VAL(haystack))[k]` seems "non idiomatic" in this file, I would rather see
`ZSTR_VAL(haystack) + ZSTR_LEN(haystack) - ZSTR_LEN(needle)` and get rid of `k` entirely
(but let's see what nikic thinks)

wkhudgins92 replied:
I'm alright with changing it, I don't do a lot of C so I had to play around some to figure that one out. More than happy to use a more concise or idiomatic way.

wkhudgins92 commented on ext/standard/tests/strings/str_starts_with.phpt:
```diff
@@ -0,0 +1,47 @@
+var_dump(str_starts_with($testStr, "beginning"));                                                                             
+var_dump(str_starts_with($testStr, "Beginning"));                                                                             
+var_dump(str_starts_with($testStr, "eginning"));  
+var_dump(str_starts_with($testStr, $testStr."a"));  
```
Good catch!

wkhudgins92 commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,45 @@
+var_dump(str_ends_with($testStr, "End"));
+var_dump(str_ends_with($testStr, "end"));
+var_dump(str_ends_with($testStr, "en"));
+var_dump(str_ends_with($testStr, $testStr."a"));
```
I added such a case but kept this one

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -10,15 +10,15 @@ var_dump(str_ends_with($testStr, "End"));
 var_dump(str_ends_with($testStr, "end"));
 var_dump(str_ends_with($testStr, "en"));
 var_dump(str_ends_with($testStr, $testStr."a"));
+var_dump(str_ends_with("a".$testStr, $testStr));
```
Sorry, that's not what I had in mind (like this it's basically testing the same thing as the very first test), I meant `($testStr, "a".$testStr)` (and expect `false`).
Moreover, if you want to also keep `($testStr, $testStr."a")` here, then you should add `($testStr, "a".$testStr)` in the other file too ;)

whudgins830 replied:
Haha I took it the way you intended it at first, then I told myself you might've meant the other way xD

guilliamxavier commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,42 @@ PHP_FUNCTION(str_contains)
+/* {{{ proto bool str_starts_with(string haystack, string needle)
+   Checks if haystack starts with needle */
+PHP_FUNCTION(str_starts_with) {
+    zend_string *haystack, *needle;
```
I just noticed that you used 4-spaces indentation, but this project uses tabs for C files (cf https://github.com/php/php-src/blob/master/.editorconfig)

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,47 @@
+var_dump(str_ends_with($testStr, "End"));
+var_dump(str_ends_with($testStr, "end"));
+var_dump(str_ends_with($testStr, "en"));
+var_dump(str_ends_with($testStr, $testStr."a"));
+var_dump(str_ends_with($testStr, "a".$testStr));
```
```suggestion
var_dump(str_ends_with($testStr, $testStr));
var_dump(str_ends_with($testStr, $testStr.$testStr));
```
(i.e. add a test that a string ends with itself, and merge both `$testStr."a"` and `"a".$testStr` into a single `$testStr.$testStr`)

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,47 @@
+bool(true)
+bool(false)
+bool(false)
+bool(false)
+bool(false)
```
```suggestion
bool(true)
bool(false)
```

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,47 @@
+var_dump(str_ends_with($testStr, "\x00"));
+var_dump(str_ends_with("\x00", ""));
+var_dump(str_ends_with("\x00", "\x00"));
+var_dump(str_ends_with("a", "\x00"));
```
```suggestion
```
(as said previously, this one doesn't add anything over `($testStr, "\x00")` above)

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -0,0 +1,47 @@
+bool(false)
+bool(true)
+bool(true)
+bool(false)
```
```suggestion
```

guilliamxavier commented on ext/standard/tests/strings/str_starts_with.phpt:
```diff
@@ -0,0 +1,43 @@
+var_dump(str_starts_with($testStr, "beginning"));
+var_dump(str_starts_with($testStr, "Beginning"));
+var_dump(str_starts_with($testStr, "eginning"));
+var_dump(str_starts_with($testStr, $testStr."a"));
```
```suggestion
var_dump(str_starts_with($testStr, $testStr));
var_dump(str_starts_with($testStr, $testStr.$testStr));
```
(like in the other file)

guilliamxavier commented on ext/standard/tests/strings/str_starts_with.phpt:
```diff
@@ -0,0 +1,43 @@
+bool(true)
+bool(false)
+bool(false)
+bool(false)
```
```suggestion
bool(true)
bool(false)
```

wkhudgins92 commented:
Thanks for the suggestions, @guilliamxavier I incorporated them into the tests!

guilliamxavier commented on ext/standard/tests/strings/str_ends_with.phpt:
```diff
@@ -39,7 +38,6 @@ bool(true)
 bool(true)
 bool(false)
 bool(true)
-bool(true)
```
this was the wrong line to delete (but OK you "fixed" it in the next commit)

guilliamxavier commented on ext/standard/tests/strings/str_starts_with.phpt:
```diff
@@ -9,7 +9,7 @@ $testStr = "beginningMiddleEnd";
 var_dump(str_starts_with($testStr, "Beginning"));
 var_dump(str_starts_with($testStr, "eginning"));
-var_dump(str_starts_with($testStr, $testStr."a"));
+var_dump(str_starts_with($testStr, $testStr));
```
still missing
```php
var_dump(str_starts_with($testStr, $testStr.$testStr));
```
after this line (and a corresponding
```
bool(false)
```
below)

TysonAndre commented:
@wkhudgins92  - I'd suggest that you squash the 22 commits in https://github.com/php/php-src/pull/5300/commits so that it can get merged

- Other RFCs are generally merged as a single commit, e.g. 1668ad7cb17213e67e42994e0c6911e302a3c3c5
- It makes sense to have each commit work as a change on its own - some of these changes only make sense in combination with other changes (e.g. "changed spaces to tabs")

nikic commented on ext/standard/string.c:
```diff
@@ -1866,6 +1866,44 @@ PHP_FUNCTION(str_contains)
+PHP_FUNCTION(str_ends_with) 
+{
+	zend_string *haystack, *needle;
+	int k;
```
This needs to use size_t, otherwise it may truncate.
